### PR TITLE
Julia 1.0 compatibility updates.

### PR DESCRIPTION
**Pull request by doorisajar**
The main updates to be compatible with Julia 1.0 are:

- Can't add methods to `ifelse` anymore, so it had to be renamed (to `ifelse_conditional`)
- Change to how special character marcos such as `@?` have to be defined
- Broadcasting tweaks (need to wrap in `Ref` now)
- `indmin` became `argmin`
- `readstring` was removed
- a case where `Expr.args` might contain a `QuoteNode` was handled
- Minor changes around the usage of `Pkg` and `Test`

For tests to pass, this also depends on a minor 1.0-related update to `Cbc` to resolve the deprecation of `warn_once`, which [I have a separate PR for](https://github.com/JuliaOpt/Cbc.jl/pull/89). Otherwise, everything appears to be working.

### CODE REVIEW

File: test/REQUIRE [hunk 1]
```diff
@@ -1,2 +1,4 @@
 Cbc
+LinearAlgebra
```
Comment (tkoolen) [inline test/REQUIRE]: This should not be in `REQUIRE`.

Comment (doorisajar) [reply test/REQUIRE]: True! Done.

File: appveyor.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 environment:
   matrix:
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.6/julia-0.6-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.6/julia-0.6-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/1.0/julia-1.0-latest-win32.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/1.0/julia-1.0-latest-win64.exe"
```
Comment (blegat) [inline appveyor.yml]: The appveyor script can be simplified now. You can just copy paste the one in the README of:
https://github.com/JuliaCI/Appveyor.jl

File: .travis.yml [hunk 3]
```diff
@@ -26,9 +26,13 @@ branches:
 #before_script: # homebrew for mac
 #  - if [ $TRAVIS_OS_NAME = osx ]; then brew install gcc; fi
 
-## uncomment the following lines to override the default test script
-#script:
-#  - julia -e 'Pkg.clone(pwd()); Pkg.build("ConditionalJuMP"); Pkg.test("ConditionalJuMP"; coverage=true)'
+before_script:
+  - export PATH=$HOME/.local/bin:$PATH
```
Comment (rdeits) [inline .travis.yml]: Why is this necessary?

File: .travis.yml [hunk 4]
```diff
@@ -26,9 +26,13 @@ branches:
...
+
+script:
+  - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
+  - julia --check-bounds=yes -e 'using Pkg; Pkg.clone(pwd()); Pkg.build("ConditionalJuMP"); Pkg.test("ConditionalJuMP"; coverage=true)'
+  #- julia -e 'Pkg.clone(pwd()); Pkg.build("ConditionalJuMP"); Pkg.test("ConditionalJuMP"; coverage=true)'
```
Comment (rdeits) [inline .travis.yml]: Since we're just running the default test script, we can delete this whole section instead of un-commenting it.

File: .travis.yml [hunk 5]
```diff
@@ -26,9 +26,13 @@ branches:
...
 after_success:
   # push coverage results to Codecov
-  - julia -e 'cd(Pkg.dir("ConditionalJuMP")); Pkg.add("Coverage"); using Coverage; Codecov.submit(Codecov.process_folder())'
+  - julia -e 'using Pkg; cd(Pkg.dir("ConditionalJuMP")); Pkg.add("Coverage"); using Coverage; Codecov.submit(Codecov.process_folder())'
```
Comment (rdeits) [inline .travis.yml]: `Pkg.dir()` no longer exists, but it's also no longer necessary for coverage submission. This can just become: 

```
- julia -e 'import Pkg; Pkg.add("Coverage"); using Coverage; Codecov.submit(Codecov.process_folder())'
```

File: README.md [hunk 6]
```diff
@@ -93,7 +93,7 @@ solve(m)
 
 ## Disjunctions
 
-This package also provides the `@ifelse` macro to create simple `if` statements which work both on numbers and on optimization variables. For example, let's write a simple update function:
+This package also provides the `@ifelse_conditional` macro to create simple `if` statements which work both on numbers and on optimization variables. For example, let's write a simple update function:
```
Comment (rdeits) [inline README.md]: I'm fine with changing the function name to `ifelse_conditional`, but I would rather keep the macro as `@ifelse`. I just checked, and it seems that there's no problem having a macro with that name.

File: src/ConditionalJuMP.jl [hunk 7]
```diff
@@ -32,7 +34,7 @@ end
 
 """
 Naive O(N^2) simplification. Slower for very large expressions, but allocates
-no memory and is solidly faster for expressions with < 100 variables. 
+no memory and is solidly faster for expressions with < 100 variables.
 """
 function simplify_inplace!(e::JuMP.GenericAffExpr{T, Variable}) where T
```
Comment (rdeits) [inline src/ConditionalJuMP.jl:39]: It's possible that this function is no longer needed, since JuMP now already stores its terms as OrderedDicts without duplicates. But as long as the code is still working, we don't need to worry about it for this PR.

File: test/runtests.jl [hunk 8]
```diff
@@ -29,11 +32,11 @@ end
         ]
         for (test_name, num_vars_total, num_vars_in_expr, num_trials) in test_cases
             @testset "$(test_name)" begin
-                srand(42)
                 m = Model()
+                seed = Random.seed!(42)
                 @variable m q[1:num_vars_total]
                 for i in 1:num_trials
-                    N = rand(1:num_vars_in_expr)
+                    N = rand(seed, 1:num_vars_in_expr)
```
Comment (rdeits) [inline test/runtests.jl]: Is passing `seed` to this particular `rand()` call necessary? Since `Random.seed!()` already sets the global RNG, I don't think it affects the behavior.

**Review by rdeits — COMMENTED** [review]
Thank you so much! This is great. I just have a few minor comments, and I'd like to get that Cbc.jl PR merged and tagged before we merge this. Thanks!

File: test/runtests.jl (hunk 8 above)
Comment (doorisajar) [reply test/runtests.jl]: Good point.

File: appveyor.yml (hunk 2 above)
Comment (ianfiske) [reply appveyor.yml]: Thanks @blegat -- I just pushed that change.

File: .travis.yml (hunk 3 above)
Comment (ianfiske) [reply .travis.yml]: Not sure, just copied it off of another project we have working.  Will try removing and see if tests pass.

File: .travis.yml (hunk 5 above)
Comment (ianfiske) [reply .travis.yml]: Thank you @rdeits -- pushed that change.

File: README.md (hunk 6 above)
Comment (doorisajar) [reply README.md]: Tested as well, looks good.

File: .travis.yml (hunk 4 above)
Comment (ianfiske) [reply .travis.yml]: We were getting build errors with the default test script.  I changed to a specified test command and the errors resolved.  I just reverted back to the default test script to double check and will post those errors if they re-emerge.

Comment (ianfiske) [reply .travis.yml]: Ghost in the machine!  The default test script is working fine now.  The only test failure is at the expected place that depends on merging/releasing https://github.com/JuliaOpt/Cbc.jl/pull/89.

**Comment by doorisajar** [comment]
OK, I believe everything to date is addressed pending the `Cbc.jl` PR being merged (which is the cause of the test failure). The `Cbc.jl` PR is approved, but not merged yet.

**Comment by doorisajar** [comment]
`Cbc.jl` PR is now merged!

**Comment by rdeits** [comment]
Sorry for the delay! I've just restarted the Travis CI builds, so let's see if they pass. If the CI passes, then I'm happy to merge this.

**Comment by rdeits** [comment]
Merged. Thanks everybody!
